resources/js/mp-history/main.tsx

```diff
@@ -0,0 +1,222 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the GNU Affero General Public License v3.0.
+// See the LICENCE file in the repository root for full licence text.
+
+import HeaderV4 from 'components/header-v4';
+import LegacyMatchEventJson from 'interfaces/legacy-match-event-json';
+import LegacyMatchJson from 'interfaces/legacy-match-json';
+import UserJson from 'interfaces/user-json';
+import { route } from 'laroute';
+import { keyBy, dropRightWhile } from 'lodash';
+import * as React from 'react';
+import { classWithModifiers } from 'utils/css';
+import Content from './content';
+
+interface LegacyMultiplayerHistoryJson {
+  current_game_id?: number;
+  events: LegacyMatchEventJson[];
+  first_event_id: number;
+  latest_event_id: number;
+  match: LegacyMatchJson;
+  users: UserJson[];
+}
+
+interface Props {
+  events: LegacyMultiplayerHistoryJson;
+}
+
+interface State {
+  currentGameId?: number;
+  events: LegacyMatchEventJson[];
+  latestEventId: number;
+  loadingNext: boolean;
+  loadingPrevious: boolean;
+  match: LegacyMatchJson;
+  users: Partial<Record<number, UserJson>>;
+}
+
+const fetchLimit = 100;
+const maximumEvents = 500;
+const refreshTimeout = 10000;
+
+export default class Main extends React.Component<Props, State> {
+  private readonly timeouts: Partial<Record<string, number>> = {};
+
+  constructor(props: Props) {
+    super(props);
+
+    this.state = {
+      currentGameId: props.events.current_game_id,
+      events: props.events.events,
+      latestEventId: props.events.latest_event_id,
+      loadingNext: false,
+      loadingPrevious: false,
+      match: props.events.match,
+      users: keyBy(props.events.users, 'id'),
+    };
+  }
+
+  componentDidMount() {
+    this.delayedAutoload();
+  }
+
+  componentWillUnmount() {
+    for (const timeout of Object.values(this.timeouts)) {
+      clearTimeout(timeout);
+    }
+  }
+
+  render() {
+    return (
+      <>
+        <HeaderV4 theme='mp-history' />
+        <div className={classWithModifiers('osu-page', ['generic'])}>
+          <Content
+            currentGameId={this.state.currentGameId}
+            events={this.state.events}
+            hasNext={this.hasNext}
+            hasPrevious={this.hasPrevious}
+            isAutoloading={this.isAutoloading}
+            loadNext={this.loadNext}
+            loadPrevious={this.loadPrevious}
+            loadingNext={this.state.loadingNext}
+            loadingPrevious={this.state.loadingPrevious}
+            match={this.state.match}
+            users={this.state.users} />
+        </div>
+      </>
+    );
+  }
+
+  private autoload() {
+    if (!this.isAutoloading) {
+      return;
+    }
+
+    this.loadNext();
+  }
+
+  private delayedAutoload() {
+    this.timeouts.autoload = setTimeout(() => this.autoload(), refreshTimeout);
+  }
+
+  private get hasLatest(): boolean {
```

Comment:
oh yeah, getters should be before constructor

(the linter for it is currently in a weird state)